bluefog/torch/optimizers.py

```diff
@@ -462,6 +463,343 @@ def step(self, closure=None):
+            state['acc_delta'] = torch.zeros_like(p.data, memory_format=torch.preserve_format)
+
+        square_avg, acc_delta = state['square_avg'], state['acc_delta']
+        rho, eps = param_group['rho'], param_group['eps']
+
+        state['step'] += 1
+
+        if param_group['weight_decay'] != 0:
+            grad = grad.add(p.data, alpha=param_group['weight_decay'])
+
+        square_avg.mul_(rho).addcmul_(grad, grad, value=1 - rho)
+        std = square_avg.add(eps).sqrt_()
+        delta = acc_delta.add(eps).sqrt_().div_(std).mul_(grad)
+        p.data.add_(delta, alpha=-param_group['lr'])
+        acc_delta.mul_(rho).addcmul_(delta, delta, value=1 - rho)
+
+    def _neighbor_allreduce_data_async(self, p):
+        name = self._parameter_names.get(p)
+        handle = bf.neighbor_allreduce_nonblocking(p.data, name=name, self_weight=self.self_weight,
+                                                   neighbor_weights=self.neighbor_weights,
+                                                   send_neighbors=self.send_neighbors,
+                                                   enable_topo_check=self.enable_topo_check)
+        return handle
+
+    def _hierarchical_neighbor_allreduce_data_async(self, p):
+        name = self._parameter_names.get(p)
+        handle = bf.hierarchical_neighbor_allreduce_nonblocking(
+            p.data, name=name, self_weight=self.self_weight,
+            neighbor_machine_weights=self.neighbor_machine_weights,
+            send_neighbor_machines=self.send_neighbor_machines,
+            enable_topo_check=self.enable_topo_check)
+        return handle
+
+    def _allreduce_data_async(self, p):
+        name = self._parameter_names.get(p)
+        handle = bf.allreduce_nonblocking(p.data, average=True, name=name)
+        return handle
+
+    def turn_on_timeline(self):
+        handles = _register_timeline(
+            self, self._models, self._parameter_names, self._communication_type.value)
+        self._timeline_hook_handles.extend(handles)
+        self._use_timeline = True
+
+    def turn_off_timeline(self):
+        for hook in self._timeline_hook_handles:
+            hook.remove()
+        self._timeline_hook_handles.clear()
+        self._use_timeline = False
+
+    def synchronize(self):
+        with torch.no_grad():
+            for p, handle in self._handles.items():
+                if handle is not None:
+                    output = bf.synchronize(handle)
+                    p.set_(output)
+                self._reduce_delay[p] = self._num_steps_per_communication
+        self._handles.clear()
+
+        self._synchronized = True
+
+    def step(self, closure=None):
+         # Step 0 is called for parameter initialization after parameter broadcast
+        if bf.size() > 1 and self._handles:
+            loss = None
+            if closure is not None:
+                loss = closure()
+            self.synchronize()
+            self._synchronized = False
```

Comment:
Not sure this variable is meaningful here, it always stays at False. (I know it was set to True in synchronize(), but immediately turns to False)  Shall we do it in the hook instead, or just remove it if it is no use?

Reply:
Add a todo here. I still think about how to make the sync in ATC more meaningful.